Add ES93BloomFilterStoredFieldsFormat for efficient field existence checks

fcofdez opened the pull request:
Introduces a new stored fields format that builds a Bloom filter for a specific field to enable fast existence checks without storing the field itself. This delegates storage of all other fields to another StoredFieldsFormat while maintaining the ability to quickly determine if a document might contain the target field.

### Missing parts
- Merge support for the bloom filter
- Dynamic bloom filter sizing
- More extensive testing

fcofdez committed:
Add ES93BloomFilterStoredFieldsFormat for efficient field existence checks

Introduces a new stored fields format that builds a Bloom filter for a
specific field to enable fast existence checks without storing the field
itself. This delegates storage of all other fields to another
StoredFieldsFormat while maintaining the ability to quickly determine if
a document might contain the target field.

tlrx commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterHashFunctions.java:
```diff
@@ -0,0 +1,194 @@
+     *      Apache Hive Murmer3</a>
+     * @since 1.13
+     */
+    public static final class MurmurHash3 {
```
This is just moved code, right?

fcofdez replied:
Yes, I just wanted to use the same hash function as the other bloom filter is using.

fcofdez replied:
Yes, I was a bit confused about why we had the same hash algorithm implemented multiple times in seemingly different ways... maybe the version that I moved is a bit more performant?

tlrx commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java:
```diff
@@ -0,0 +1,539 @@
+
+    static {
+        // Precompute powers of two (2^1 to 2^26) for efficient modulo operations using bitwise AND.
+        // We start from 2^1 (2 bits) and go up to 2^26 (67,108,864 bits / 8,388,608 bytes = 8 MB)
```
> // We start from 2^1 (2 bits) 

I think it start computing the power of two from 2 ⁰ in the following loop?

tlrx commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java:
```diff
@@ -0,0 +1,539 @@
+        int bloomFilterSizeInBits = 0;
+        // Find the closest power of 2 that fits the required size
+        for (int powerOfTwoBitSetSize : powerOfTwoBitSetSizes) {
+            if (powerOfTwoBitSetSize <= (Math.multiplyExact(bloomFilterSize.getBytes(), Byte.SIZE))) {
```
nit: we can compute the multiplication before the loop.

fcofdez replied:
That's a great simplification, I've tackled it in https://github.com/elastic/elasticsearch/pull/137331/commits/a1af7b966a245fdceacabf1e723402564fc6ec1a

tlrx commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java:
```diff
@@ -0,0 +1,539 @@
+                bloomFilterSizeInBits = powerOfTwoBitSetSize;
+            }
+        }
+        assert bloomFilterSizeInBits > 0;
```
```suggestion
        assert bloomFilterSizeInBits > 0 : bloomFilterSizeInBits;
```

fcofdez replied:
Tackled in https://github.com/elastic/elasticsearch/pull/137331/commits/a1af7b966a245fdceacabf1e723402564fc6ec1a

tlrx commented on server/src/test/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormatTests.java:
```diff
@@ -0,0 +1,162 @@
+                    BigArrays.NON_RECYCLING_INSTANCE,
+                    "",
+                    TestUtil.getDefaultCodec().storedFieldsFormat(),
+                    ByteSizeValue.ofKb(2),
```
Would it make sense to randomize this a bit?

tlrx commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java:
```diff
@@ -0,0 +1,539 @@
+
+        @Override
+        public void writeField(FieldInfo info, BytesRef value) throws IOException {
+            if (info.getName().equals(bloomFilterFieldName)) {
```
I wonder if we could lookup the FieldInfo.number for the `_id` field once, and then only compare `int`s?

tlrx replied:
Or maybe just add (bloomFilterFieldInfo != null && bloomFilterFieldInfo.number == info.number) ?

fcofdez replied:
Good point, tackled in https://github.com/elastic/elasticsearch/pull/137331/commits/a1af7b966a245fdceacabf1e723402564fc6ec1a

tlrx commented on server/src/test/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormatTests.java:
```diff
@@ -0,0 +1,162 @@
+                    "",
+                    TestUtil.getDefaultCodec().storedFieldsFormat(),
+                    ByteSizeValue.ofKb(2),
+                    IdFieldMapper.NAME
```
I wonder if we could have a better coverage if we were randomly using some field names that are generated in BaseStoredFieldsFormatTestCase?

tlrx reviewed (commented):
Looks great! I left many comments but most of them are nits.

I think this is ready for review, we'll improve the format step by step.

fcofdez committed:
Review comments

Skip skipping the bloom filter field for all data types
Compute the bloom filter size in a simpler way

fcofdez committed:
Add more randomness

fcofdez commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java:
```diff
@@ -0,0 +1,539 @@
+            String bloomFilterFieldName,
+            StoredFieldsWriter delegateWriter
+        ) throws IOException {
+            assert isPowerOfTwo(bloomFilterSizeInBits) : "Bloom filter size is not a power of 2";
```
Tackled in https://github.com/elastic/elasticsearch/pull/137331/commits/a1af7b966a245fdceacabf1e723402564fc6ec1a

fcofdez commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java:
```diff
@@ -0,0 +1,539 @@
+            if (info.getName().equals(bloomFilterFieldName)) {
+                addToBloomFilter(info, value);
+            } else {
+                delegateWriter.writeField(info, value);
```
That's a good point, what I did was just skipping storing the value for the rest of types other than binary, otherwise we would add unnecessary complexity. wdyt?

tlrx replied:
I think it depends on how this bloom stored field format will be registered. If it's registered only for the `_id` field through (an adjusted version of) `PerFieldFormatSupplier`, what I think we'll do, then I prefer to have all field behave the same as we know other types of fields won't be called.

fcofdez committed:
Update docs/changelog/137331.yaml

tlrx commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java:
```diff
@@ -0,0 +1,539 @@
+
+        private void finishBloomFilterStoredFormat() throws IOException {
+            BloomFilterMetadata bloomFilterMetadata = null;
+            if (bloomFilterFieldInfo != null) {
```
Makes sense, thanks!

tlrx reviewed (approved):
LGTM

fcofdez committed:
Improve test a bit

fcofdez committed:
Merge branch 'bloom-filter-stored-fields-format' of github.com:fcofdez/elasticsearch into bloom-filter-stored-fields-format

fcofdez committed:
Merge remote-tracking branch 'origin/main' into bloom-filter-stored-fields-format

kkrik-es reviewed (approved):
@martijnvg knows this part better for sure.

martijnvg reviewed (approved):
LGTM 👍

fcofdez committed:
Add more assertions

fcofdez committed:
Start version from 0

fcofdez committed:
Use toIntExact

fcofdez committed:
Add assertion for bloom filter name

fcofdez committed:
Extract interface

fcofdez committed:
Merge remote-tracking branch 'origin/main' into bloom-filter-stored-fields-format

fcofdez commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java:
```diff
@@ -0,0 +1,553 @@
+    public static final String STORED_FIELDS_BLOOM_FILTER_FORMAT_NAME = "ES93BloomFilterStoredFieldsFormat";
+    public static final String STORED_FIELDS_BLOOM_FILTER_EXTENSION = "sfbf";
+    public static final String STORED_FIELDS_METADATA_BLOOM_FILTER_EXTENSION = "sfbfm";
+    private static final int VERSION_START = 1;
```
Addressed in https://github.com/elastic/elasticsearch/pull/137331/commits/605c93de9cf0df375d6af7d37c48c66155b47740
